### ISSUE TITLE
Offer to install type stub packages for users

### ISSUE THREAD

**Issue by gramster** [microsoft/pylance-release#3702]
For some set of blessed packages that we know have well maintained corresponding stub packages on PyPI, we should detect that the user is using the package and offer to install the stub package.

**Comment by rchiodo** [comment]
@luabud I'm going to try doing the auto-download version of this to see how it goes. Meaning there wouldn't be any GUI changes for this, so it shouldn't need a spec? Do you think the auto-download into a cached location is enough of a spec for this?

**Comment by rchiodo** [comment]
Going to start with just django-stubs. See if that can dynamically download, then add more later.

**Comment by rchiodo** [comment]
Yes, I think that would be possible. I'll do that as a second step.

**Comment by rchiodo** [comment]
This is partially implemented. Just for Django. Enabling an experiment to see if we get any complaints or problems with the idea. 

Next steps:
- Download sqlalchemy stubs for SqlAlchemy 1.x

**Comment by debonte** [comment]
> Download sqlalchemy stubs for SqlAlchemy 1.x

Didn't we decide not to do this because the same stubs would be used for SQLAlchemy 2.x and degrade the experience there?

Could we download them to a workspace-specific location? The `typings` directory?

**Comment by rchiodo** [comment]
> Could we download them to a workspace-specific location? The typings directory?

That might work but I don't like that we'd be exposing our implementation to the user. Maybe instead we could have a per workspace cache for certain scenarios. That cache could still be written in the pylance\dist folder, but there'd be a subfolder per workspace.

**Comment by rchiodo** [comment]
And this would only be for certain cases. Most would use the 'global' cache location.

**Comment by rchiodo** [comment]
Having second thoughts about this. Downloading is messy. And downloading when the package is already there is usually too late. User might start writing code for a package long before they actually run the code. 

We'd have to download when they:
- Type the name of the package somewhere? What if they're opening a file with it already in it
- We see the name of the package in a file in the workspace? What if this makes 10 different downloads happen and meanwhile the completions don't work. 

Did some investigation on the size of the bundled stubs. 

We bundle 12 different packages right now and it comes out to 1.8MB on disk. My guess is nobody would notice until we got into 10s of MBs.

So assuming an average compressed size is .15MB per package, we could theoretically bundle a 100 best known stubs before anybody really cared. 

Bundled stubs are better for a number of reasons:
- They work when the user doesn't have internet
- They get updated by us and there's no danger of making requests that the user may not want to have happen
- They let the user have completions for packages they haven't installed yet

Moving back to triage. Maybe the better solution here is to just bundle more stuff.

**Comment by rchiodo** [comment]
The use case of downloading stubs per workspace so that SQLAlchemy 1 can work seems like a one off. If we really need to support that we could special case using SQLAlchemay stubs only when SQLAlchemy is not 2.0 or higher.

**Comment by rchiodo** [comment]
Switching this to be what @gramster originally asked for. Have an action to just install type stubs in the current envrionment.

Whether or not this uses pip/poetry/conda complicates things, but maybe for now we can just have it work for venv environments. 

It would also have an 'allowed' list of stubs we know work. Like SQLAlchemy 1.0 stubs.

**Comment by heejaechang** [comment]
This issue has been fixed in prerelease version 2026.3.100, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20263100-24-August-2026-PreRelease)
